File: examples/telegraph_examples.ts
/**
 * Telegraph Examples - TypeScript/Deno
 * Demonstrates all major Telegraph use cases
 */

import TelegraphPublisher from "../scripts/telegraph_publisher.ts";
import {
  CodeBlockFormatter,
  TableFormatter,
  MetricsFormatter,
  SubagentOutputIntegration,
  MediaHandler,
  TelegraphCliHelper,
} from "../scripts/telegraph_integration.ts";

/**
 * Example 1: Simple Markdown publication
 */
async function example1SimplMarkdown(): Promise<string> {
  console.log("\n" + "=".repeat(60));
  console.log("Example 1: Simple Markdown Publication");
  console.log("=".repeat(60));

  // Create publisher and account
  const publisher = new TelegraphPublisher();
  const token = await publisher.createAccount(
    "deno_example1",
    "OpenClaw"
  );
  console.log(`✓ Account created with token: ${token.substring(0, 30)}...`);

  // Markdown content
  const markdownContent = `
# Hello Telegraph!

This is a **simple** Markdown document published to Telegraph.

## Features

- Easy to use
- Supports *markdown* formatting
- Fast publication

[Visit GitHub](https://github.com)
`;

  // Publish
  const url = await publisher.publishMarkdown(
    "Simple Markdown Example (Deno)",
    markdownContent,
    "OpenClaw"
  );

  console.log(`✓ Published to: ${url}`);
  return url;
}

/**
 * Example 2: Code documentation with syntax highlighting
 */
async function example2CodeDocumentation(): Promise<string> {
  console.log("\n" + "=".repeat(60));
  console.log("Example 2: Code Documentation with Syntax Highlighting");
  console.log("=".repeat(60));

  const publisher = new TelegraphPublisher();
  await publisher.createAccount("deno_example2", "OpenClaw");

  // TypeScript code example
  const typeScriptCode = `
interface TelegraphConfig {
  apiUrl: string;
  maxRetries: number;
}

class TelegraphPublisher {
  async publishMarkdown(
    title: string,
    content: string
  ): Promise<string> {
    // Implementation here
    return url;
  }
}
`;

  // Format code block
  const formattedCode = CodeBlockFormatter.formatCodeBlock(
    typeScriptCode,
    "typescript"
  );

  const htmlContent = `
<h1>TypeScript Telegraph Integration</h1>
<p>A complete example of using Telegraph with TypeScript:</p>
${formattedCode}
<p><strong>Features:</strong></p>
<ul>
    <li>Full type safety</li>
    <li>Modern async/await</li>
    <li>Deno and Node.js support</li>
</ul>
`;

  // Publish
  const url = await publisher.publishHtml(
    "TypeScript Code Documentation",
    htmlContent,
    "Code Reviewer"
  );

  console.log(`✓ Published to: ${url}`);
  return url;
}

/**
 * Example 3: Status report with GFM (GitHub Flavored Markdown)
 */
async function example3StatusReport(): Promise<string> {
  console.log("\n" + "=".repeat(60));
  console.log("Example 3: Status Report with GFM");
  console.log("=".repeat(60));

  const publisher = new TelegraphPublisher();
  await publisher.createAccount("deno_example3", "OpenClaw");

  // Report data
  const reportData: Record<string, any> = {
    summary: "Weekly development status - All on track",
    metrics: {
      "Completed Tasks": "24",
      "In Progress": "8",
      "Pending": "12",
      "Build Success": "98.5%",
      "Test Coverage": "92%",
    },
    sections: {
      "Development": "Completed core API endpoints and database migrations.",
      "Testing": "All unit tests passing, integration tests 95% complete.",
      "DevOps": "CI/CD pipeline optimized, deployment time reduced by 30%.",
    },
  };

  // Format as status report
  const htmlContent = MetricsFormatter.formatStatusReport(reportData);

  // Publish
  const url = await publisher.publishHtml(
    `Development Status Report - ${new Date().toISOString().split("T")[0]}`,
    htmlContent,
    "Project Lead"
  );

  console.log(`✓ Published to: ${url}`);
  return url;
}

/**
 * Example 4: Blog post publishing
 */
async function example4BlogPost(): Promise<string> {
  console.log("\n" + "=".repeat(60));
  console.log("Example 4: Blog Post Publishing");
  console.log("=".repeat(60));

  const publisher = new TelegraphPublisher();
  await publisher.createAccount("deno_blog_example", "OpenClaw Blog");

  const blogContent = `
# Telegraph + Deno: A Powerful Publishing Combination

Telegraph is an excellent platform for publishing content,
and Deno provides a modern runtime for TypeScript.

## Why Telegraph?

Telegraph offers several advantages:

1. **Simplicity** - Clean API, minimal overhead
2. **Speed** - Content published instantly
3. **Reliability** - Built on robust infrastructure
4. **Flexibility** - HTML and Markdown support

## Getting Started with Deno

Here's a quick example:

\`\`\`typescript
import TelegraphPublisher from "./telegraph_publisher.ts";

const publisher = new TelegraphPublisher();
await publisher.createAccount();

const url = await publisher.publishMarkdown(
  "My First Post",
  "Hello from Deno!"
);
console.log(\`Published: \${url}\`);
\`\`\`

## Key Features

- Full TypeScript support
- Zero dependencies
- Works in both Deno and Node.js
- Automatic retries with exponential backoff
- Media upload support

## Conclusion

Telegraph + Deno is the ideal combination for fast, reliable content publishing.

**Happy publishing!**
`;

  const url = await publisher.publishMarkdown(
    "Telegraph + Deno: Publishing Made Easy",
    blogContent,
    "OpenClaw Team"
  );

  console.log(`✓ Published to: ${url}`);
  return url;
}

/**
 * Example 5: Media embedding
 */
async function example5MediaEmbedding(): Promise<string> {
  console.log("\n" + "=".repeat(60));
  console.log("Example 5: Media Embedding");
  console.log("=".repeat(60));

  const publisher = new TelegraphPublisher();
  await publisher.createAccount("deno_media_example", "OpenClaw");

  const mediaHandler = new MediaHandler(publisher);

  // Create content with media
  const htmlContent = `
<h1>Media Embedding Example</h1>
<p>This page demonstrates media embedding in Telegraph:</p>

<h2>Embedded Video</h2>
<p>You can embed YouTube videos using iframe:</p>
${mediaHandler.createVideoEmbed("dQw4w9WgXcQ")}

<h2>Benefits of Rich Media</h2>
<ul>
    <li>Enhanced engagement</li>
    <li>Professional presentation</li>
    <li>Interactive content</li>
</ul>

<h2>Gallery of Options</h2>
<p>You can embed:</p>
<ul>
    <li>Images (JPG, PNG, GIF)</li>
    <li>Videos (YouTube, Vimeo)</li>
    <li>Iframes (custom content)</li>
</ul>
`;

  // Publish
  const url = await publisher.publishHtml(
    "Media Embedding Example",
    htmlContent,
    "Media Producer"
  );

  console.log(`✓ Published to: ${url}`);
  return url;
}

/**
 * Example 6: Table formatting
 */
async function example6TableFormatting(): Promise<string> {
  console.log("\n" + "=".repeat(60));
  console.log("Example 6: Table Formatting");
  console.log("=".repeat(60));

  const publisher = new TelegraphPublisher();
  await publisher.createAccount("deno_table_example", "OpenClaw");

  // Markdown table
  const markdownTable = `
| Feature | Status | Progress |
|---------|--------|----------|
| Core API | ✓ Done | 100% |
| Frontend | ⚠ WIP | 75% |
| Mobile | ◯ Planned | 0% |
| Docs | ✓ Done | 100% |
`;

  // Convert to HTML
  const htmlTable = TableFormatter.markdownTableToHtml(markdownTable);

  const htmlContent = `
<h1>Product Roadmap</h1>
<p>Current progress across all components:</p>
${htmlTable}
<p style="font-style: italic; margin-top: 20px;">
    Last updated: ${new Date().toLocaleString()}
</p>
`;

  const url = await publisher.publishHtml(
    "Product Roadmap Status",
    htmlContent,
    "Product Manager"
  );

  console.log(`✓ Published to: ${url}`);
  return url;
}

/**
 * Example 7: Subagent output integration
 */
async function example7SubagentIntegration(): Promise<string> {
  console.log("\n" + "=".repeat(60));
  console.log("Example 7: Subagent Output Integration");
  console.log("=".repeat(60));

  const publisher = new TelegraphPublisher();
  await publisher.createAccount("deno_subagent_example", "OpenClaw");

  // Simulate subagent output
  const subagentOutput = `
# Performance Analysis
## Request Summary
Task: Analyze and optimize database queries

## Key Findings

1. **Query Optimization**
   - Reduced average query time: 450ms → 120ms
   - Added strategic indexes
   - Implemented query caching

2. **Results**
   - Overall performance improved 73%
   - Memory usage down 25%
   - API response time halved

## Implementation

\`\`\`typescript
// Added database indexing
CREATE INDEX idx_user_email ON users(email);
CREATE INDEX idx_post_created ON posts(created_at);

// Query optimization
async function getUser(id: string) {
  return cache.get(\`user:\${id}\`) || 
    db.query('SELECT * FROM users WHERE id = ?', [id]);
}
\`\`\`

## Recommendations
- Continue monitoring slow queries
- Schedule weekly performance reviews
- Plan next optimization phase

## Status
✓ Complete - Ready for production deployment
`;

  // Use integration helper
  const integration = new SubagentOutputIntegration(publisher);
  const url = await integration.processSubagentOutput(
    subagentOutput,
    "Performance Analysis Report",
    "Database Optimization"
  );

  console.log(`✓ Published to: ${url}`);
  return url;
}

/**
 * Run all examples
 */
async function runAllExamples(): Promise<void> {
  console.log("\n" + "=".repeat(70));
  console.log("Telegraph Publishing Examples (TypeScript/Deno)");
  console.log("=".repeat(70));

  const examples: Array<{
    name: string;
    fn: () => Promise<string>;
  }> = [
    { name: "Simple Markdown", fn: example1SimplMarkdown },
    { name: "Code Documentation", fn: example2CodeDocumentation },
    { name: "Status Report", fn: example3StatusReport },
    { name: "Blog Post", fn: example4BlogPost },
    { name: "Media Embedding", fn: example5MediaEmbedding },
    { name: "Table Formatting", fn: example6TableFormatting },
    { name: "Subagent Integration", fn: example7SubagentIntegration },
  ];

  const results: Array<{
    name: string;
    result: string;
    status: string;
  }> = [];

  for (const { name, fn } of examples) {
    try {
      const url = await fn();
      results.push({ name, result: url, status: "✓" });
    } catch (error) {
      console.error(`✗ Error: ${error}`);
      results.push({ name, result: String(error), status: "✗" });
    }
  }

  // Summary
  console.log("\n" + "=".repeat(70));
  console.log("Example Results Summary");
  console.log("=".repeat(70));

  for (const { name, result, status } of results) {
    console.log(`${status} ${name}: ${result.substring(0, 80)}`);
  }

  console.log("\n✓ All examples completed!");
}

// Run examples if this is the main module
if (import.meta.main) {
  runAllExamples();
}

export {
  example1SimplMarkdown,
  example2CodeDocumentation,
  example3StatusReport,
  example4BlogPost,
  example5MediaEmbedding,
  example6TableFormatting,
  example7SubagentIntegration,
  runAllExamples,
};

File: examples/telegraph_examples.py
#!/usr/bin/env python3
"""
Telegraph Examples for OpenClaw
Demonstrates all major use cases
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'scripts'))

from telegraph_publisher import TelegraphPublisher
from telegraph_integration import (
    CodeBlockFormatter, TableFormatter, MetricsFormatter,
    SubagentOutputIntegration, TelegraphCliHelper
)
from datetime import datetime


def example_1_simple_markdown():
    """Example 1: Simple Markdown publication"""
    print("\n" + "="*60)
    print("Example 1: Simple Markdown Publication")
    print("="*60)
    
    # Create publisher and account
    publisher = TelegraphPublisher()
    token = publisher.create_account(
        short_name="openclaw_example1",
        author_name="OpenClaw"
    )
    print(f"✓ Account created with token: {token[:30]}...")
    
    # Markdown content
    markdown_content = """
    # Hello Telegraph!
    
    This is a **simple** Markdown document published to Telegraph.
    
    ## Features
    
    - Easy to use
    - Supports *markdown* formatting
    - Fast publication
    
    [Visit GitHub](https://github.com)
    """
    
    # Publish
    url = publisher.publish_markdown(
        title="Simple Markdown Example",
        content=markdown_content,
        author="OpenClaw"
    )
    
    print(f"✓ Published to: {url}")
    return url


def example_2_code_documentation():
    """Example 2: Code documentation with syntax highlighting"""
    print("\n" + "="*60)
    print("Example 2: Code Documentation with Syntax Highlighting")
    print("="*60)
    
    publisher = TelegraphPublisher()
    publisher.create_account(
        short_name="openclaw_example2",
        author_name="OpenClaw"
    )
    
    # Code example
    python_code = """
def fibonacci(n):
    '''Generate Fibonacci sequence up to n'''
    if n <= 0:
        return []
    elif n == 1:
        return [0]
    
    sequence = [0, 1]
    while sequence[-1] + sequence[-2] <= n:
        sequence.append(sequence[-1] + sequence[-2])
    return sequence

# Usage
print(fibonacci(100))
"""
    
    # Format code block
    formatted_code = CodeBlockFormatter.format_code_block(python_code, "python")
    
    html_content = f"""
    <h1>Fibonacci Implementation</h1>
    <p>A simple Python implementation of the Fibonacci sequence:</p>
    {formatted_code}
    <p><strong>Features:</strong></p>
    <ul>
        <li>Efficient algorithm</li>
        <li>Readable code</li>
        <li>Proper documentation</li>
    </ul>
    """
    
    # Publish
    url = publisher.publish_html(
        title="Code Documentation Example",
        content=html_content,
        author="Code Reviewer"
    )
    
    print(f"✓ Published to: {url}")
    return url


def example_3_status_report():
    """Example 3: Status report with metrics"""
    print("\n" + "="*60)
    print("Example 3: Status Report with Metrics")
    print("="*60)
    
    publisher = TelegraphPublisher()
    publisher.create_account(
        short_name="openclaw_example3",
        author_name="OpenClaw"
    )
    
    # Report data
    report_data = {
        "summary": "Daily system status report - All systems operational",
        "metrics": {
            "Uptime": "99.98%",
            "Active Tasks": "24",
            "Completed": "156",
            "Pending": "12",
            "API Latency": "145ms",
            "Error Rate": "0.02%"
        },
        "sections": {
            "System Status": "All systems are running normally. No alerts.",
            "Performance": "Average response time improved by 12% this week.",
            "Deployments": "3 successful deployments, 0 rollbacks"
        }
    }
    
    # Format as status report
    html_content = MetricsFormatter.format_status_report(report_data)
    
    # Publish
    url = publisher.publish_html(
        title=f"Daily Status Report - {datetime.now().strftime('%Y-%m-%d')}",
        content=html_content,
        author="System Monitor"
    )
    
    print(f"✓ Published to: {url}")
    return url


def example_4_blog_post():
    """Example 4: Blog post publishing"""
    print("\n" + "="*60)
    print("Example 4: Blog Post Publishing")
    print("="*60)
    
    publisher = TelegraphPublisher()
    publisher.create_account(
        short_name="openclaw_blog_example",
        author_name="OpenClaw Blog"
    )
    
    blog_content = """
# Getting Started with Telegraph
## A Guide to Publishing Content

Telegraph is a powerful platform for publishing content quickly and easily.
In this post, we'll explore how to use Telegraph with OpenClaw.

### Why Telegraph?

Telegraph offers several advantages:

1. **Speed** - Publish content instantly
2. **Simplicity** - Minimal API, maximum functionality
3. **Reliability** - Built on solid infrastructure
4. **Flexibility** - Supports HTML, Markdown, and more

### Getting Started

Here's a quick example:

```python
from telegraph_publisher import TelegraphPublisher

publisher = TelegraphPublisher()
publisher.create_account()

url = publisher.publish_markdown(
    title="My First Post",
    content="Hello, Telegraph!"
)
print(f"Published: {url}")
```

### Best Practices

- Keep titles concise and descriptive
- Use proper formatting for readability
- Include relevant metadata
- Test before publishing

### Conclusion

Telegraph combined with OpenClaw provides a powerful platform
for automated content publishing and distribution.

**Happy publishing!**
"""
    
    url = publisher.publish_markdown(
        title="Getting Started with Telegraph",
        content=blog_content,
        author="OpenClaw Team"
    )
    
    print(f"✓ Published to: {url}")
    return url


def example_5_media_embedding():
    """Example 5: Media embedding (images and videos)"""
    print("\n" + "="*60)
    print("Example 5: Media Embedding")
    print("="*60)
    
    publisher = TelegraphPublisher()
    publisher.create_account(
        short_name="openclaw_media_example",
        author_name="OpenClaw"
    )
    
    # Create a sample image for upload (1x1 PNG)
    import base64
    from pathlib import Path
    
    # 1x1 transparent PNG
    png_data = base64.b64decode(
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg=="
    )
    
    # Save sample image
    sample_image = Path("/tmp/sample.png")
    sample_image.write_bytes(png_data)
    
    try:
        # Upload media
        media_url = publisher.upload_media(str(sample_image))
        print(f"✓ Media uploaded: {media_url}")
        
        # Create content with embedded media
        html_content = f"""
        <h1>Media Embedding Example</h1>
        <p>This page demonstrates media embedding in Telegraph:</p>
        
        <h2>Embedded Image</h2>
        <img src="{media_url}" alt="Sample Image" style="max-width: 400px;">
        
        <h2>YouTube Embedding</h2>
        <p>You can also embed YouTube videos using iframe:</p>
        <iframe width="400" height="300" 
                src="https://www.youtube.com/embed/dQw4w9WgXcQ" 
                frameborder="0" allowfullscreen></iframe>
        
        <h2>Benefits</h2>
        <ul>
            <li>Rich multimedia content</li>
            <li>Enhanced engagement</li>
            <li>Professional presentation</li>
        </ul>
        """
        
        # Publish
        url = publisher.publish_html(
            title="Media Embedding Example",
            content=html_content,
            author="Media Producer"
        )
        
        print(f"✓ Published to: {url}")
        return url
    
    finally:
        # Cleanup
        if sample_image.exists():
            sample_image.unlink()


def example_6_table_formatting():
    """Example 6: Table formatting"""
    print("\n" + "="*60)
    print("Example 6: Table Formatting")
    print("="*60)
    
    publisher = TelegraphPublisher()
    publisher.create_account(
        short_name="openclaw_table_example",
        author_name="OpenClaw"
    )
    
    # Markdown table
    markdown_table = """
    | Feature | Status | Progress |
    |---------|--------|----------|
    | Core API | ✓ Complete | 100% |
    | Dashboard | ⚠ In Progress | 75% |
    | Mobile App | ◯ Planned | 0% |
    | Documentation | ✓ Complete | 100% |
    """
    
    # Convert to HTML
    html_table = TableFormatter.markdown_table_to_html(markdown_table)
    
    html_content = f"""
    <h1>Project Status Table</h1>
    <p>Current project progress across all components:</p>
    {html_table}
    <p style="font-style: italic; margin-top: 20px;">
        Last updated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S EDT')}
    </p>
    """
    
    url = publisher.publish_html(
        title="Project Status Report",
        content=html_content,
        author="Project Manager"
    )
    
    print(f"✓ Published to: {url}")
    return url


def example_7_subagent_integration():
    """Example 7: Subagent output integration"""
    print("\n" + "="*60)
    print("Example 7: Subagent Output Integration")
    print("="*60)
    
    publisher = TelegraphPublisher()
    publisher.create_account(
        short_name="openclaw_subagent_example",
        author_name="OpenClaw"
    )
    
    # Simulate subagent output
    subagent_output = """
# Analysis Report
## Request Summary
Task: Analyze system performance metrics

## Key Findings

1. **Performance Improvement**
   - Average response time: 245ms → 180ms
   - Throughput increase: 23%
   
2. **Resource Optimization**
   - Memory usage down 15%
   - CPU utilization stable at 42%

## Code Changes
```python
def optimize_cache(cache_size):
    # Increased cache efficiency
    return cache_size * 1.3
```

## Recommendations
- Continue monitoring metrics
- Plan next optimization cycle
- Document changes in wiki

## Status
✓ Complete - Ready for deployment
"""
    
    # Use integration helper
    integration = SubagentOutputIntegration(publisher)
    url = integration.process_subagent_output(
        subagent_output,
        title="Subagent Analysis Report",
        author="Performance Analyst"
    )
    
    print(f"✓ Published to: {url}")
    return url


def run_all_examples():
    """Run all examples"""
    print("\n" + "="*70)
    print("Telegraph Publishing Examples for OpenClaw")
    print("="*70)
    
    examples = [
        ("Simple Markdown", example_1_simple_markdown),
        ("Code Documentation", example_2_code_documentation),
        ("Status Report", example_3_status_report),
        ("Blog Post", example_4_blog_post),
        ("Media Embedding", example_5_media_embedding),
        ("Table Formatting", example_6_table_formatting),
        ("Subagent Integration", example_7_subagent_integration),
    ]
    
    results = []
    
    for name, example_func in examples:
        try:
            url = example_func()
            results.append((name, url, "✓"))
        except Exception as e:
            print(f"✗ Error: {e}")
            results.append((name, str(e), "✗"))
    
    # Summary
    print("\n" + "="*70)
    print("Example Results Summary")
    print("="*70)
    
    for name, result, status in results:
        print(f"{status} {name}: {result[:80]}")
    
    print("\n✓ All examples completed!")


if __name__ == "__main__":
    # Run all examples
    run_all_examples()
